modules/tts/tts-adapters/tts-adapter-cereproc/src/main/java/org/daisy/pipeline/tts/cereproc/impl/CereProcEngine.java

```diff
@@ -202,4 +213,133 @@ public TTSResource allocateThreadResources() throws SynthesisException, Interrup
 		}
 		return result;
 	}
+
+	/**
+	 * Function to transform ssml to String and also perform Regex rules and char substitutions.
+	 *
+	 * @param
+	 * @return
+	 * @throws SaxonApiException
+	 */
+		public String transformSSML(XdmNode ssmlIn, Voice v) throws SynthesisException, SaxonApiException, XMLStreamException {
+
+		    List<XdmItem> ssmlProcessed = new ArrayList<>();
+
+			Configuration conf = ssmlIn.getUnderlyingNode().getConfiguration();
+
+			XMLStreamReader reader = new SaxonInputValue(ssmlIn, conf).asXMLStreamReader();
+			XMLStreamWriter writer = new SaxonOutputValue(item -> {
+				if (item instanceof XdmNode) {
+					ssmlProcessed.add(item);
+				} else {
+					throw new RuntimeException(); // should not happen
+				}
+			}, conf).asXMLStreamWriter();
+
+			performSubstitutionRules(reader, writer, v.getLocale().get().getLanguage());
+
+			if ( ssmlProcessed.size() != 1) {
+				throw new RuntimeException("Something went wrong");
+			}
+			if (!(ssmlProcessed.get(0) instanceof XdmNode)) {
+				throw new RuntimeException("Incorrect type");
+			}
+
+			XdmNode ssmlOut = (XdmNode) ssmlProcessed.get(0);
+
+			CompiledStylesheet styleSheet = this.getStyleSheetFromConfig(conf);
+			return styleSheet.newTransformer().transformToString(ssmlOut);
+		}
+
+	/**
+	 * Cache stylesheets so they dont need to be compiled every time.
+	 * @param conf
+	 * @return
+	 * @throws SynthesisException
+	 */
+	private CompiledStylesheet getStyleSheetFromConfig(Configuration conf) throws SynthesisException {
+		XslTransformCompiler xslCompiler = new XslTransformCompiler(conf);
+		CompiledStylesheet styleSheet;
+		styleSheet = this.mStyleSheets.get(conf);
+		if (styleSheet == null) {
+			try {
+				styleSheet = xslCompiler.compileStylesheet(CereProcEngine.class.getResourceAsStream("/transform-ssml.xsl"));
```

Comment:
This could be done in `CereProcService`, or at least the `AbstractTTSService.getSSMLxslTransformerURL()` method could be used get the URL. This has the benefit that it will also work with OSGi.